working/macros/example/data_class.dart

```diff
@@ -5,225 +5,235 @@ import '../api/code.dart';
-              '${field.type.toCode()}${field.type.isNullable ? '' : '?'} '
-              '${field.name}'),
-      ];
-      var args = [
-        for (var field in allFields)
-          NamedArgumentCode.fromString(
-              '${field.name}: ${field.name} ?? this.${field.name}'),
-      ];
-      builder.declareInClass(DeclarationCode.fromParts([
-        clazz.instantiate().toCode(),
-        ' copyWith({',
-        ...namedParams.joinAsCode(', '),
-        ',})',
-        // TODO: We assume this constructor exists, but should check
-        '=> ', clazz.instantiate().toCode(), '(',
-        ...args.joinAsCode(', '),
-        ', );',
-      ]));
-    });
+  Future<void> buildDeclarationsForClass(
+      ClassDeclaration clazz, ClassDeclarationBuilder builder) async {
+    var methods = await builder.methodsOf(clazz);
+    if (methods.any((c) => c.name == 'copyWith')) {
+      throw ArgumentError(
+          'Cannot generate a copyWith method because one already exists');
+    }
+    var allFields = await clazz.allFields(builder).toList();
+    var namedParams = [
+      for (var field in allFields)
+        ParameterCode.fromString(
+            '${field.type.toCode()}${field.type.isNullable ? '' : '?'} '
+            '${field.name}'),
+    ];
+    var args = [
+      for (var field in allFields)
+        NamedArgumentCode.fromString(
+            '${field.name}: ${field.name} ?? this.${field.name}'),
+    ];
+    builder.declareInClass(DeclarationCode.fromParts([
+      clazz.instantiate().toCode(),
+      ' copyWith({',
+      ...namedParams.joinAsCode(', '),
+      ',})',
+      // TODO: We assume this constructor exists, but should check
+      '=> ', clazz.instantiate().toCode(), '(',
+      ...args.joinAsCode(', '),
+      ', );',
+    ]));
   }
 }
 
 const hashCode = _HashCode();
 
-class _HashCode implements ClassMacro {
+class _HashCode implements ClassDeclarationsMacro, ClassDefinitionMacro {
   const _HashCode();
 
   @override
-  void visitClass(ClassDeclaration clazz, ClassContext context) {
-    context.buildDeclarations((builder) {
-      builder.declareInClass(DeclarationCode.fromString('''
+  void buildDeclarationsForClass(
+      ClassDeclaration clazz, ClassDeclarationBuilder builder) {
+    builder.declareInClass(DeclarationCode.fromString('''
 @override
 external int get hashCode;'''));
-    });
+  }
 
-    context.buildDefinitions((builder) async {
-      await builder.buildMethod('hashCode', (builder) async {
-        var hashCodeExprs = [
-          await for (var field in clazz.allFields(builder))
-            ExpressionCode.fromString('${field.name}.hashCode')
-        ].joinAsCode(' ^ ');
-        builder.augment(FunctionBodyCode.fromParts([
-          ' => ',
-          ...hashCodeExprs,
-          ';',
-        ]));
-      });
+  @override
+  Future<void> buildDefinitionForClass(
+      ClassDeclaration clazz, ClassDefinitionBuilder builder) async {
+    await builder.buildMethod('hashCode', (builder) async {
```

Comment:
Since the API is explicitly phased, we don't need the callbacks anymore, do we? At this point, it seems like all it does is give you a FunctionDefinitionBuilder for the right method. But a simpler API for that is something like:

```dart
@override
Future<void> buildDefinitionForClass(
    ClassDeclaration clazz, ClassDefinitionBuilder builder) async {
  var functionBuilder = builder.buildMethod('hashCode');
  var hashCodeExprs = [
    await for (var field in clazz.allFields(builder))
      ExpressionCode.fromString('${field.name}.hashCode')
  ].joinAsCode(' ^ ');
  functionBuilder.augment(FunctionBodyCode.fromParts([
    ' => ',
    ...hashCodeExprs,
    ';',
  ]));
}
```

Reply:
Yes true, ill switch to that.